bench: replace embedded raw block with configurable block generator

l0rinc opened the pull request:
**Problem:** The [hardcoded benchmark block 413567](https://mempool.space/block/413567) has several issues:
- embeds ~1MB of binary data in the repository (we can't remove it from history, but we can from `HEAD`)
- mined in 2016, predates SegWit v0, Taproot, and modern script types
- benchmarks cannot test different transaction patterns or script mixes
- updating to a newer block (e.g., [#32457](https://github.com/bitcoin/bitcoin/pull/32457)) would still add binary bloat and would remain inflexible

**Fix:** Added configurable `bench/block_generator.{h,cpp}` that builds fully valid blocks at runtime, providing blocks or raw serialized streams.

**Reproducer:** You can run all affected tests and benchmarks with a sanity check:
```bash
rm -rf build && \
cmake -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_BENCH=ON && \
cmake --build build -j$(nproc) && \
build/bin/test_bitcoin -t block_generator_deterministic_seeded_output,block_generator_serialization_roundtrip,block_generator_seed_perturbation,block_generator_multiple_seed_sanity && \
build/bin/bench_bitcoin -sanity-check -filter='DeserializeBlockTest|CheckBlockTest|LoadExternalBlockFile|WriteBlockBench|ReadBlockBench|ReadRawBlockBench|BlockToJsonVerbosity1|BlockToJsonVerbosity2|BlockToJsonVerbosity3|BlockToJsonVerboseWrite|HexStrBench'
```
---

This supersedes #32457. New benchmark scenarios can vary transaction and script mixes without adding another embedded block.

yuvicc commented:
Concept ACK

This one is pretty useful as I can test with up-to date test_data.

Raimo33 commented:
Concept ACK for more realistic benchmarks

hodlinator reviewed (commented):
Concept ACK b5b57cdf396d8945612fcb2d51e530475db0d87d

**Strong problem A-C-K - It's quite bad that we are only benchmarking against a pre-segwit block.**

The general direction is promising, I'd mainly prefer slightly more realistic transactions.

Also, since removing the old block data doesn't shrink the size of the repo, maybe we could keep it around together with some of the older benchmarks?

Suggestions branch: https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/32554_suggestions
(Corresponds pretty much to the inline comments).

l0rinc committed:
bench: route legacy fixture through helper APIs

Bench files loaded `block413567.raw` directly, so moving off the fixture required touching each benchmark's serialization setup.
Add shared helper entry points backed by the current fixture and switch benchmark callsites to those helpers.
This keeps behavior stable while moving setup to `regtest`, where the follow-up generated-block path can find a valid nonce quickly.

Co-authored-by: hodlinator <hodlinator@users.noreply.github.com>
Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

l0rinc committed:
bench: switch helper to generated blocks and recipes

The helper migration still returned one fixed historical fixture block, which kept benchmark coverage narrow and blocked script-mix experiments.
Replace the fixture-backed helper with runtime block generation and `ScriptRecipe` presets so callsites can tune output mixes.
Add unit tests that lock in deterministic seed behavior and helper round-trip validity.
Link `bench_bitcoin` with `secp256k1` so benchmark code can include `<secp256k1.h>` and use canonical pubkey tag constants.

Co-authored-by: hodlinator <hodlinator@users.noreply.github.com>

l0rinc commented on src/bench/block_generator.h:
```diff
@@ -0,0 +1,79 @@
+    const ScriptRecipe& = kWitness,
+    const uint256& seed = {}
+);
+CBlock GetBlock(
```
Done, extracted this to a separate commit where we're wrapping the previous behavior first to separate the introduction of the generator from applying it.

l0rinc commented on src/bench/block_generator.cpp:
```diff
@@ -0,0 +1,177 @@
+    block.nVersion = 1 + rng.randrange<int>(VERSIONBITS_LAST_OLD_BLOCK_VERSION);
+    block.nTime = params.GenesisBlock().nTime;
+    block.hashPrevBlock.SetNull();
+    block.nBits = UintToArith256(params.GetConsensus().powLimit).GetCompact(); // lowest difficulty
```
Added genesis as `hashPrevBlock`, thanks.

l0rinc commented on src/bench/readwriteblock.cpp:
```diff
@@ -19,32 +19,26 @@
+    const auto test_block{benchmark::GetBlock(params)};
     const auto& expected_hash{test_block.GetHash()};
-    const auto& pos{blockman.WriteBlock(test_block, 413'567)};
+    const auto& pos{blockman.WriteBlock(test_block, /*nHeight=*/1)};
```
we have a genesis now so kept the 1 - thanks!

l0rinc committed:
bench: remove embedded `block413567.raw` fixture

After benchmarks switched to helper-generated data, the embedded `block413567.raw` fixture and its cmake raw-data wiring were no longer needed.

Co-authored-by: hodlinator <hodlinator@users.noreply.github.com>

l0rinc commented on src/bench/block_generator.cpp:
```diff
@@ -0,0 +1,177 @@
+            probability -= p;
+        }
+        const auto raw{rng.randbytes<uint8_t>(1 + rng.randrange(100))};
+        return CScript(raw.begin(), raw.end());
```
> I think it would be better to require that ScriptRecipe probabilities add up to ~1.0 and not generate random scripts as a fallback.

Good idea, added `random_script_prob` to make this explicit.

l0rinc reviewed (commented):
Thanks a lot for the excellent review (I see you started it in December already).
I have applied almost every suggestion in one shape or form - let's iterate on a few remaining ones, let's first see if there's an appetite for removing the pre-segwit block and having a configurable generator.

Split the change up to smaller commits, first migrating the existing benchmarks to a common interface, followed by introducing the new generator (and the deletion as a separate commit since that huge binary is breaking the diff tools for me)
Added you as coauthor.
